Source/Parallelization/WarpXComm.cpp

```diff
@@ -909,21 +956,65 @@ WarpX::SyncRho ()
     if (!rho_fp[0]) return;
     const int ncomp = rho_fp[0]->nComp();
 
-    // Restrict fine patch onto the coarse patch,
-    // before summing the guard cells of the fine patch
-    for (int lev = 1; lev <= finest_level; ++lev)
+    std::unique_ptr<MultiFab> mf_comm; // for communication between levels
+    for (int lev = finest_level; lev >= 0; --lev)
     {
-        rho_cp[lev]->setVal(0.0);
-        const IntVect& refinement_ratio = refRatio(lev-1);
-        ablastr::coarsen::average::Coarsen(*rho_cp[lev], *rho_fp[lev], refinement_ratio );
-    }
+        if (lev < finest_level)
+        {
+            auto const& period = Geom(lev).periodicity();
+
+            // On a coarse level, the data in mf_comm comes from the
+            // fine level. They are unfiltered and uncommuncatied. We
+            // need to add it to the current level.
```

Comment:
I think we should improve this inline comment too, as the one above:
```suggestion
            // On a coarse level, the data in mf_comm comes from the
            // coarse patch of the fine level. They are unfiltered and uncommunicated.
            // We need to add it to the fine patch of the current level.
```